File: DSAlgorithms/Leetcode/c++/linked_list/25-reverse_nodes_in_k_groups.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
 class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
       if(k <= 1 || head == nullptr) 
            return head;
       ListNode* current = head;
       ListNode* prev = nullptr;
       while(true)
       {
            ListNode* last = prev;
            ListNode* new_end = current;
            //reverse between left and right
            ListNode* next = current->next;
            ListNode* temp = nullptr;
            int count = 0;
            temp = current;
            while(temp != nullptr ) 
            {
                temp = temp->next;
                count++;  
            }           
            if(count < k)
                break;
            for(int i = 0; current!= nullptr && i < k ; i++)
            {
                current->next = prev;
                prev = current;
                current = next;
                if(next != nullptr)
                    next = next->next;
            }
            if(last != nullptr)
                last->next = prev;
            else
                head = prev;
        
            new_end->next = current;
            if(current == nullptr)
                break; 
            prev = new_end;
       }
        return head;
    }
};

File: dsa_algo/leetcode/ts/src/25-reverse_nodes_in_k_groups.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

const reverseKGroup = (head: ListNode | null, k: number): ListNode | null =>{
    const dummy: ListNode = new ListNode();
    dummy.next = head;
    let prev: ListNode = dummy;
    let current: ListNode = dummy.next;
    let temp:ListNode = null;
    let count = k;
    while(current !== null)
    {
        if(count > 1)
        {
            temp = prev.next;
            prev.next = current.next;
            current.next = current.next.next;
            prev.next.next = temp;
            count--;
        }
        else
        {
            prev = current;
            current = current.next;
            count = k
            let end: ListNode = current;
            for(let i = 0; i < k; i++)
            {
                if(end === null)
                    return dummy.next;
                end = end.next;
            }
        }
    }
    return dummy.next;
};